File: Backend/src/main/java/com/sih/gem/service/DocumentService.java
package com.sih.gem.service;

import com.sih.gem.entity.BidderDocument;
import com.sih.gem.entity.DocStage;
import com.sih.gem.repository.BidderDocumentRepository;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Service
public class DocumentService {

    private final BidderDocumentRepository documentRepository;
    private final DocumentProcessor documentProcessor;

    @Value("${app.upload.dir}")
    private String uploadDir;

    public DocumentService(BidderDocumentRepository documentRepository,
                           DocumentProcessor documentProcessor) {
        this.documentRepository = documentRepository;
        this.documentProcessor = documentProcessor;
    }

    public List<BidderDocument> getDocumentsForBid(String bidId) {
        return documentRepository.findByBidIdOrderByUploadedAtDesc(bidId);
    }

    public List<BidderDocument> getUserDocuments(String email) {
        if (email == null || email.isBlank()) {
            return List.of();
        }
        return documentRepository.findByUploadedByOrderByUploadedAtDesc(email);
    }

    public DocumentDownload getDownload(Long documentId, String requester, boolean officer) {
        BidderDocument document = documentRepository.findById(documentId)
                .orElseThrow(() -> new IllegalArgumentException("Document was not found."));
        if (!officer && (document.getUploadedBy() == null || !document.getUploadedBy().equalsIgnoreCase(requester))) {
            throw new org.springframework.security.access.AccessDeniedException("You cannot download another bidder's document.");
        }
        try {
            Path path = Paths.get(document.getStoredPath()).toAbsolutePath().normalize();
            if (!Files.isRegularFile(path)) throw new IllegalArgumentException("The document file is unavailable.");
            return new DocumentDownload(new UrlResource(path.toUri()), document.getFilename());
        } catch (IOException | RuntimeException ex) {
            if (ex instanceof org.springframework.security.access.AccessDeniedException accessDenied) throw accessDenied;
            throw new IllegalArgumentException("The document file is unavailable.");
        }
    }

    public BidderDocument uploadDocument(String bidId, MultipartFile file, String docType, String uploadedBy) {
        try {
            if (bidId == null || !bidId.matches("[A-Za-z0-9][A-Za-z0-9._-]{0,63}")) {
                throw new IllegalArgumentException("Invalid bid identifier.");
            }
            if (file == null || file.isEmpty()) {
                throw new IllegalArgumentException("The uploaded document is empty.");
            }
            // 1. Save file to disk
            String originalName = file.getOriginalFilename() == null ? "document" : file.getOriginalFilename();
            String ext = getExtension(originalName);
            if (!"pdf".equals(ext)) {
                throw new IllegalArgumentException("Only PDF documents are supported by the current AI pipeline.");
            }
            try (var input = file.getInputStream()) {
                byte[] signature = input.readNBytes(5);
                if (signature.length < 5 || !"%PDF-".equals(new String(signature, java.nio.charset.StandardCharsets.US_ASCII))) {
                    throw new IllegalArgumentException("The uploaded file is not a valid PDF document.");
                }
            }
            // Keep the original basename for downstream AI provenance while
            // isolating each upload in a unique directory.
            String safeOriginalName = originalName.replaceAll("[^A-Za-z0-9._-]", "_");
            Path dir = Paths.get(uploadDir, bidId, UUID.randomUUID().toString());
            Files.createDirectories(dir);
            Path target = dir.resolve(safeOriginalName);
            file.transferTo(target.toAbsolutePath());

            // 2. Build document entity
            BidderDocument doc = BidderDocument.builder()
                    .bidId(bidId)
                    .filename(originalName)
                    .docType(docType == null || docType.isBlank() ? "Bidder Certificate" : docType)
                    .fileFormat(ext.toUpperCase())
                    .fileSize(formatSize(file.getSize()))
                    .uploadStatus("UPLOADED")
                    .processingStatus("PROCESSING")
                    .uploadedAt(LocalDateTime.now())
                    .uploadedBy(uploadedBy)
                    .storedPath(target.toAbsolutePath().toString())
                    .progressPercentage(10)
                    .stages(new ArrayList<>())
                    .build();

            doc.getStages().add(stage("Uploaded", true, "File uploaded successfully"));
            doc.getStages().add(stage("Content Detected", true, ext.toUpperCase() + " format recognized"));
            documentRepository.save(doc);

            // 3. Process document synchronously / pipeline
            documentProcessor.process(doc);

            return doc;
        } catch (IOException e) {
            throw new RuntimeException("Failed to store document: " + e.getMessage(), e);
        }
    }

    private DocStage stage(String name, boolean completed, String details) {
        return DocStage.builder()
                .name(name)
                .completed(completed)
                .timestamp(LocalDateTime.now().format(DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm")))
                .details(details)
                .build();
    }

    private String getExtension(String filename) {
        int idx = filename.lastIndexOf('.');
        return idx >= 0 ? filename.substring(idx + 1).toLowerCase() : "pdf";
    }

    private String formatSize(long bytes) {
        double mb = bytes / (1024.0 * 1024.0);
        return String.format("%.1f MB", mb);
    }

    public record DocumentDownload(Resource resource, String filename) {}
}

File: Frontend/Frontend/src/services/documentService.ts
import { api, USE_MOCK } from './api';
import { mockUserBidderDocuments, mockBids } from '../data/mockData';
import type { BidderDocument } from '../types';

let userDocsState: BidderDocument[] = mockUserBidderDocuments.map((document) => ({
  ...document,
  bidId: document.bidId || mockBids[0]?.id || ''
}));

export const documentService = {
  async getDocuments(bidId: string): Promise<BidderDocument[]> {
    if (USE_MOCK) {
      await new Promise((r) => setTimeout(r, 300));
      return userDocsState.filter((document) => document.bidId === bidId);
    }
    const response = await api.get<BidderDocument[]>(`/bids/${bidId}/documents`);
    return response.data;
  },

  async getUserDocuments(): Promise<BidderDocument[]> {
    if (USE_MOCK) {
      await new Promise((r) => setTimeout(r, 200));
      const currentUserEmail = localStorage.getItem('gem_user_email') || '';
      return userDocsState.filter((d) => d.uploadedBy === currentUserEmail);
    }
    const response = await api.get<BidderDocument[]>('/user/documents');
    return response.data;
  },

  async uploadTenderDocument(bidId: string, file: File): Promise<BidderDocument> {
    if (USE_MOCK) {
      await new Promise((r) => setTimeout(r, 800));
      const ext = file.name.split('.').pop()?.toUpperCase() || 'PDF';
      const currentUserEmail = localStorage.getItem('gem_user_email') || '';
      const uploaded: BidderDocument = {
        id: `DOC-TND-${Date.now()}`,
        bidId,
        filename: file.name,
        docType: "Tender RFP Specification",
        fileFormat: ext,
        fileSize: `${(file.size / (1024 * 1024)).toFixed(1)} MB`,
        uploadStatus: "UPLOADED",
        processingStatus: "PROCESSING",
        uploadedAt: new Date().toISOString(),
        unitLabel: "Pages",
        progressPercentage: 10,
        uploadedBy: currentUserEmail
      };
      userDocsState = [uploaded, ...userDocsState.filter((document) => !(document.bidId === bidId && document.docType.toLowerCase().includes('tender')) )];
      return uploaded;
    }
    const formData = new FormData();
    formData.append('file', file);
    const response = await api.post<BidderDocument>(`/bids/${bidId}/tender-document`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return response.data;
  },

  async uploadBidderDocument(bidId: string, file: File, docType: string): Promise<BidderDocument> {
    if (USE_MOCK) {
      await new Promise((r) => setTimeout(r, 700));
      const ext = file.name.split('.').pop()?.toUpperCase() || 'PDF';
      const currentUserEmail = localStorage.getItem('gem_user_email') || '';

      const newDoc: BidderDocument = {
        id: `DOC-BID-${Date.now()}`,
        bidId,
        filename: file.name,
        docType: docType || "Bidder Certificate",
        fileFormat: ext,
        fileSize: `${(file.size / (1024 * 1024)).toFixed(1)} MB`,
        uploadStatus: "UPLOADED",
        processingStatus: "PROCESSING",
        uploadedAt: new Date().toISOString(),
        unitLabel: ext === 'PDF' ? 'Pages' : undefined,
        progressPercentage: 10,
        uploadedBy: currentUserEmail,
        stages: [
          { name: "Uploaded", completed: true, timestamp: "Just now", details: "File uploaded successfully" },
          { name: "Content Detected", completed: false, details: "Awaiting processing service" },
          { name: "Content Processed", completed: false },
          { name: "Ready for Verification", completed: false }
        ]
      };

      userDocsState = [newDoc, ...userDocsState];
      return newDoc;
    }
    const formData = new FormData();
    formData.append('file', file);
    formData.append('docType', docType);
    const response = await api.post<BidderDocument>(`/bids/${bidId}/documents`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return response.data;
  },

  async downloadDocument(documentId: string): Promise<Blob> {
    const response = await api.get<Blob>(`/documents/${encodeURIComponent(documentId)}/download`, {
      responseType: 'blob'
    });
    return response.data;
  }
};
